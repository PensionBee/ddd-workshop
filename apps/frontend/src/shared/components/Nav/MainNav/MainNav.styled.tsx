import styled, { css } from "styled-components";
import { mediaFrom } from "@shared/styling/mediaQueries";
import { theme } from "@shared/styling/theme";

export type StyledMenuProps = {
  $borderBottom?: boolean;
};

export const StyledMenu = styled.div<StyledMenuProps>`
  position: relative;
  display: flex;
  justify-content: center;
  width: 100%;
  z-index: 1;
  ${({ $borderBottom }) =>
    $borderBottom &&
    css`
      border-bottom: 1px solid ${theme.colors.grey4};
      justify-content: flex-start;
    `}
`;

export type StyledListProps = {
  $colGapLarge: number;
};

export const StyledList = styled.ul<StyledListProps>`
  column-gap: 20px;
  display: flex;
  flex-direction: row;
  background-color: #ffffff;
  flex-wrap: wrap;
  ${mediaFrom.lg} {
    column-gap: ${({ $colGapLarge }) => $colGapLarge}px;
  }
`;
