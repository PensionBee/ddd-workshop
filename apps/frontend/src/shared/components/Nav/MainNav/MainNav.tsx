import { useClickOutside } from "@shared/hooks/useClickOutside/useClickOutside";
import { useFocusOutside } from "@shared/hooks/useFocusOutside/useFocusOutside";
import { useRef, useState } from "react";
import { useArrowNavigation } from "../hooks/useArrowNavigation/useArrowNavigation";
import type { NavMenuItem } from "../Nav.types";
import type { StyledMenuProps, StyledListProps } from "./MainNav.styled";
import { StyledMenu, StyledList } from "./MainNav.styled";
import { MainNavItem } from "./MainNavItem/MainNavItem";
import { MainNavLink } from "./MainNavLink/MainNavLink";

export type MainNavProps = {
  list: NavMenuItem[];
  borderBottom?: StyledMenuProps["$borderBottom"];
  colGapLarge?: StyledListProps["$colGapLarge"];
  isTabMenu?: boolean;
};

export const MainNav: React.FC<MainNavProps> = ({
  borderBottom,
  colGapLarge = 40,
  isTabMenu,
  list,
  ...props
}) => {
  const menuRef = useRef<HTMLDivElement | null>(null);
  const [menuOpen, setMenuOpen] = useState<string | null>(null);

  useClickOutside(menuRef, () => setMenuOpen(null));
  useFocusOutside(menuRef, () => setMenuOpen(null));

  const handleToggleSubMenu = (key: NavMenuItem["children"]) => {
    return setMenuOpen((menuOpen) => (menuOpen !== key ? key : null));
  };

  const handleCloseMenu = () => {
    return setMenuOpen(null);
  };

  useArrowNavigation(menuRef, {
    direction: "horizontal",
    disabled: Boolean(menuOpen),
  });

  return (
    <StyledMenu
      data-rc="MainNav"
      ref={menuRef}
      $borderBottom={borderBottom}
      {...props}
    >
      <StyledList $colGapLarge={colGapLarge}>
        {list.map((item) => {
          if (item.hidden) return null;
          if (item.subList?.length) {
            return (
              <MainNavItem
                key={item.children}
                $isTabMenu={isTabMenu}
                $isOpen={menuOpen === item.children}
                toggleMenu={handleToggleSubMenu}
                handleCloseMenu={handleCloseMenu}
                {...item}
              />
            );
          }
          return (
            <MainNavLink
              key={item.children}
              href={item.href || ""}
              $isTabMenu={isTabMenu}
              {...item}
            >
              {item.children}
            </MainNavLink>
          );
        })}
      </StyledList>
    </StyledMenu>
  );
};
