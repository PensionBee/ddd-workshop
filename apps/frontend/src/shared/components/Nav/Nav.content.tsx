import type { NavMenuItem } from "./Nav.types";

// Example content used in tests and stories

export const VISIBLE_LINK: NavMenuItem = {
  children: "Visible Item",
  href: "/",
};
export const HIDDEN_LINK: NavMenuItem = {
  children: "Hidden Item",
  href: "/",
  hidden: true,
};
export const SUB_LIST_ITEM: NavMenuItem = {
  children: "SubList Item",
  subList: [VISIBLE_LINK],
};
export const HIDDEN_SUB_LIST_ITEM: NavMenuItem = {
  children: "Hidden SubList Item",
  subList: [VISIBLE_LINK],
  hidden: true,
};

export const LIST = [VISIBLE_LINK] satisfies NavMenuItem[];
export const LIST_WITH_HIDDEN_LINK = [HIDDEN_LINK] satisfies NavMenuItem[];
export const LIST_WITH_SUB_LIST_ITEM = [SUB_LIST_ITEM] satisfies NavMenuItem[];
export const LIST_WITH_MIXED_ITEMS = [
  VISIBLE_LINK,
  HIDDEN_LINK,
  SUB_LIST_ITEM,
  HIDDEN_SUB_LIST_ITEM,
] satisfies NavMenuItem[];
