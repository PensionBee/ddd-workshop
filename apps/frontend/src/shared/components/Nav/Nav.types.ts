import type { LargeSystemIcon } from "@shared/components/Svg";

export type NavMenuItem = {
  children: string;
  href?: string;
  subList?: NavMenuItem[];
  hidden?: boolean;
  icon?: React.ComponentProps<typeof LargeSystemIcon>["icon"];
};
