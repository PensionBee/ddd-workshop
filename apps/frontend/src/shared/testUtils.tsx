import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
