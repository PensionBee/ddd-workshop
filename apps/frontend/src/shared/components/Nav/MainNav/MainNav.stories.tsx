import type { ComponentStory, ComponentMeta } from "@storybook/react";
import { GridContainer } from "@shared/components/Grid";
import { LIST_WITH_MIXED_ITEMS } from "../Nav.content";
import { MainNav, type MainNavProps } from "./MainNav";
import { MemoryRouter } from "react-router-dom";
import { Box } from "@shared/components/Box";

export default {
  title: "Shared/Menus/Main Nav",
  component: MainNav,
  parameters: {
    viewMode: "story",
    previewTabs: {
      "storybook/docs/panel": { hidden: true },
    },
  },
} as ComponentMeta<typeof MainNav>;

const MainNavTemplate: ComponentStory<typeof MainNav> = (
  args: MainNavProps
) => {
  return (
    <MemoryRouter initialEntries={[{ pathname: "/example" }]}>
      <GridContainer>
        <Box
          d="flex"
          justify="space-between"
          align="center"
          py={{ xs: 5, lg: 15 }}
          w="100%"
        >
          <Box d="flex">
            <Box ml="auto" mr={{ xs: 20, md: 40 }}>
              <MainNav {...args} />
            </Box>
          </Box>
        </Box>
      </GridContainer>
    </MemoryRouter>
  );
};

export const Default = MainNavTemplate.bind({});
Default.args = {
  list: LIST_WITH_MIXED_ITEMS,
};
