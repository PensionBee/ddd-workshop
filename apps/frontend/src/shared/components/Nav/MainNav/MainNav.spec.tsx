import { routerRender } from "@shared/testUtils";
import { MainNav } from "./MainNav";
import { LIST, LIST_WITH_MIXED_ITEMS } from "../Nav.content";

describe("MainNav", () => {
  it("should render correctly", () => {
    const { queryByText, asFragment } = routerRender(<MainNav list={LIST} />);
    expect(queryByText(LIST[0].children)).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should not render hidden sub list items", () => {
    const { queryByText, asFragment } = routerRender(
      <MainNav list={LIST_WITH_MIXED_ITEMS} />
    );

    LIST_WITH_MIXED_ITEMS.forEach((item) => {
      if (item.hidden) {
        expect(queryByText(item.children)).toBe(null);
      } else {
        expect(queryByText(item.children)).toBeVisible();
      }
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
