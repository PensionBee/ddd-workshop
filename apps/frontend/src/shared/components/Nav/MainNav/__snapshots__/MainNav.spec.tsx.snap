// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MainNav > should not render hidden sub list items 1`] = `
<DocumentFragment>
  <div
    class="sc-bczSft dNgwqX"
    data-rc="MainNav"
  >
    <ul
      class="sc-gsnScm eQnjpn"
    >
      <li
        class="sc-bretMz dBffNv"
        data-rc="MainNavLink"
      >
        <a
          class="sc-crXcXV cxxyeD"
          href="/"
        >
          Visible Item
        </a>
      </li>
      <li
        class="sc-bretMz dBffNv"
        data-rc="MainNavItem"
        data-state="closed"
      >
        <button
          class="sc-kDDqET cHUMF"
        >
          SubList Item
          <span
            aria-hidden="true"
            class="sc-hKMtDE bwbOno sc-iqcnOu kngPEJ"
            data-rc="LargeSystemIcon.ChevronYellow"
          >
            <svg
              height="7"
              viewBox="0 0 12 7"
              width="12"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.406.252l-5.16 5.28a.874.874 0 0 0 0 1.216.827.827 0 0 0 1.189 0L6 2.076l4.565 4.672a.827.827 0 0 0 1.189 0 .874.874 0 0 0 0-1.216L6.594.252a.828.828 0 0 0-1.188 0z"
                fill="#F9C000"
                fill-rule="nonzero"
              />
            </svg>
          </span>
        </button>
        <div
          class="sc-ewahw dDeaeb"
        />
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`MainNav > should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bczSft dNgwqX"
    data-rc="MainNav"
  >
    <ul
      class="sc-gsnScm eQnjpn"
    >
      <li
        class="sc-bretMz dBffNv"
        data-rc="MainNavLink"
      >
        <a
          class="sc-crXcXV cxxyeD"
          href="/"
        >
          Visible Item
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`MainNav should not render hidden sub list items 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe fqQwmU"
    data-rc="MainNav"
  >
    <ul
      class="sc-gswNZR jdAFhG"
    >
      <li
        class="sc-bqWxrE dJvvDl"
        data-rc="MainNavLink"
      >
        <a
          class="sc-csuSiG lnZrdg"
          href="/"
        >
          Visible Item
        </a>
      </li>
      <li
        class="sc-bqWxrE dJvvDl"
        data-rc="MainNavItem"
        data-state="closed"
      >
        <button
          class="sc-kDvujY iVcEOy"
        >
          SubList Item
          <span
            aria-hidden="true"
            class="sc-hLBbgP fEYWcg sc-ipEyDJ guhhVG"
            data-rc="LargeSystemIcon.ChevronYellow"
          >
            <svg>
              chevron-yellow.svg
            </svg>
          </span>
        </button>
        <div
          class="sc-eDWCr haeOkf"
        />
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`MainNav should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe fqQwmU"
    data-rc="MainNav"
  >
    <ul
      class="sc-gswNZR jdAFhG"
    >
      <li
        class="sc-bqWxrE dJvvDl"
        data-rc="MainNavLink"
      >
        <a
          class="sc-csuSiG lnZrdg"
          href="/"
        >
          Visible Item
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
